Cluster aware routers for typed

johanandren opened the pull request:
## Purpose
Make cluster routers (and anything that uses the receptionist) aware of unreachability so that nodes that are reachable will be preferred.

## References
Refs #26355 

## Changes
The default listing from the receptionist is changed to be only reachable nodes, a separate method to access both reachable and unreachable is added for special cases.

## Background Context
This will give applications reachability awareness "for free" without changing the main API.

johanandren committed:
API draft of proposed solution

johanandren committed:
Local receptionist and docs update

johanandren committed:
Receptionist keeps track of unreachable nodes

johanandren commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/Registry.scala:
```diff
@@ -63,14 +64,34 @@ import scala.concurrent.duration.Deadline
+        }
+        builders
     }
+    (reachable.result(), all.result())
```
Mixing mutability and FP like there's no tomorrw. Sorry!

patriknw replied:
maybe better to go all in, using only mutable in a `foreach` instead of `foldLeft`?

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/ReceptionistMessages.scala:
```diff
@@ -4,11 +4,12 @@
 
 package akka.actor.typed.internal.receptionist
 
+import java.util
```
Should we be specific vs an open-ended util?

johanandren replied:
Yeah, some weird autoimport from IntelliJ I think, I'll remove it

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
````diff
@@ -81,7 +81,19 @@ abstract class ServiceKey[T] extends AbstractServiceKey { key =>
+   *   case MyServiceKey.Listing(reachable) =>
+   * ```
+   *
+   * In a non-clustered actorsystem this will always be all registered instances
````
```suggestion
   * In a non-clustered `ActorSystem` this will always be all registered instances
```

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
````diff
@@ -81,7 +81,19 @@ abstract class ServiceKey[T] extends AbstractServiceKey { key =>
+   * ```
+   *
+   * In a non-clustered actorsystem this will always be all registered instances
+   * for a service key. For a clustered environment services on nodes that has
````
```suggestion
   * for a service key. For a clustered environment services on nodes that have
```

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -81,7 +81,19 @@ abstract class ServiceKey[T] extends AbstractServiceKey { key =>
+   *
+   * In a non-clustered actorsystem this will always be all registered instances
+   * for a service key. For a clustered environment services on nodes that has
+   * been observed unreachable is not among these (note that they could have
```
```suggestion
   * been observed unreachable are not among these (note that they could have
```

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
+    def getAllServiceInstances[T](key: ServiceKey[T]): java.util.Set[ActorRef[T]]
+
+    /**
+     * Scala API: `true` only if new services was added or removed and `false` if this listing is
```
```suggestion
     * Scala API: `true` only if new services were added or removed and `false` if this listing is
```

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
+
+    /**
+     * Scala API: `true` only if new services was added or removed and `false` if this listing is
+     * only about reachability changes. Useful for subscribers that only cares about [[#allServiceInstances]].
```
```suggestion
     * only about reachability changes. Useful for subscribers only concerned with [[#allServiceInstances]].
```

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
+    def allServiceInstancesChanged: Boolean
+
+    /**
+     * JavaAPI: `true` only if new services was added or removed and `false` if this listing is
```
```suggestion
     * JavaAPI: `true` only if new services were added or removed and `false` if this listing is
```

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
+
+    /**
+     * JavaAPI: `true` only if new services was added or removed and `false` if this listing is
+     * only about reachability changes. Useful for subscribers that only cares about [[#getAllServiceInstances]].
```
same as above similar one, or
```suggestion
     * only about reachability changes. Useful for subscribers that only care about [[#getAllServiceInstances]].
```

johanandren committed:
Lang fixes from H

Co-Authored-By: Helena Edelson <helena@users.noreply.github.com>

johanandren committed:
Doc updates

johanandren committed:
Merge branch 'wip-26355-take-cluster-unreachability-into-account' of github.com:johanandren/akka into wip-26355-take-cluster-unreachability-into-account

johanandren committed:
One more doc fix

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
+     * In a non-clustered `ActorSystem` this will be `true` for all listings.
+     * For `Find` queries and the initial listing for a `Subscribe` this will always be `true`.
+     */
+    def getAllServiceInstancesChanged: Boolean
```
I think normal Java convention is to use `isX` for Boolean

patriknw replied:
I actually think we can have one `isAllServiceInstancesChanged` for both Scala and Java.

For some reason I think `allServiceInstancesChanged` can be misinterpreted as "all instances has changed"

johanandren replied:
Perhaps inverting and calling it "onlyReachableChanged" or instead?

johanandren replied:
BTW the separate method was mostly to link to the right `allServiceInstances` method from the scaladoc

patriknw replied:
I think it's good that we don't tie this to closely to reachability/unreachability. That way we could have other aspects for choosing preferred vs all in the future. E.g. other health or load metrics.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -297,6 +373,20 @@ object Receptionist extends ExtensionId[Receptionist] {
+    Listing(
+      key,
+      Set[ActorRef[T]](serviceInstances.asScala.toSeq: _*),
+      Set[ActorRef[T]](allServiceInstances.asScala.toSeq: _*),
```
better to use `scala.collection.JavaConverters` ?

johanandren replied:
Wasn't sure of the rationale for the original code being like this but thought there was a reason (tm) so that's the copy-pasta rationale

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
+
+    /**
+     * JavaAPI: `true` only if new services were added or removed and `false` if this listing is
+     * only about reachability changes. Useful for subscribers only concerned with [[#getAllServiceInstances]].
```
, which can ignore all `Listing` messages where this flag is `false`

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -340,6 +350,50 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
+            }
+          }
+          next(newRegistry)
+
```
Should we also add `UnreachableDataCenter`, which would affect all entries owned by members with the dc?
Could be in separate ticket/pr since it's not as important.

johanandren replied:
We'll get both that and an unreachable event per node right? 

Let's ticketify as future improvement.

johanandren replied:
https://github.com/akka/akka/issues/27327

patriknw commented on akka-docs/src/main/paradox/typed/routers.md:
```diff
@@ -43,8 +43,8 @@ Java
-Since the receptionist is used this means the group router is cluster aware out of the box and will pick up routees
-registered on any node in the cluster (there is currently no logic to avoid routing to unreachable nodes, see [#26355](https://github.com/akka/akka/issues/26355)).
+Since the receptionist is used this means the group router is cluster aware out of the box and will route to 
+registered actors on any node in the cluster that is reachable.
```
No changes to the routers in this PR so there must be something missing. I think we said that they would prefer the reachable, but if that is empty it would fall back to all. Better to try to send to someone than just dropping.

johanandren replied:
I didn't recall if we were 100% sure we wanted that so didn't go there yet.

johanandren replied:
Will make sure this works first then add that on top

patriknw reviewed (commented):
looking good

helena commented on akka-docs/src/main/paradox/typed/routers.md:
```diff
@@ -88,4 +88,4 @@ it will not give better performance to create more routees than there are thread
 Since the router itself is an actor and has a mailbox this means that messages are routed sequentially to the routees
 where it can be processed in parallel (depending on the available threads in the dispatcher).
-In a high throughput use cases the sequential routing could be a bottle neck. Akka Typed does not provide an optimized tool for this.
+In a high throughput use cases the sequential routing could be a bottle neck. Akka Typed does not provide an optimized tool for this.
```
Is there a notion of yet here?

johanandren replied:
You mean as in "there isn't yet"? We don't plan to introduce one to not end up with the complicated internals we have for untyped to support that.

helena commented on akka-docs/src/main/paradox/typed/actor-discovery.md:
```diff
@@ -73,8 +73,15 @@ sends a `Ping` message and when receiving the `Pong` reply it stops.
+registered actors that are reachable. The full set of actors, including unreachable ones, is available through 
+@scala[`Listing.allServiceInstances`]@java[`Listing.getAllServiceInstances`].
+
+One important difference from a local only receptions is the serialisation concerns, all messages sent to and back from 
```
```suggestion
One important difference from local only receptions are the serialization concerns, all messages sent to and back from 
```

or
```suggestion
One important difference from a local only reception are the serialization concerns, all messages sent to and back from 
```

johanandren committed:
Review feedback plus made the tests pass

 * inverted and renamed the flag that says only reachability changed for clarity
 * single flag method for Java and Scala
 * Various little fixes

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,65 @@ object Receptionist extends ExtensionId[Receptionist] {
+    def getAllServiceInstances[T](key: ServiceKey[T]): java.util.Set[ActorRef[T]]
+
+    /**
+     * Returns `false` only if new services were added or removed and `true` if this listing is
```
Should invert this sentence, it reads weird.

johanandren committed:
More lang from H

Co-Authored-By: Helena Edelson <helena@users.noreply.github.com>

johanandren committed:
Better scaladoc on onlyReachabilityChanged

johanandren committed:
Fallback to unreachable when no reachable in group router

johanandren committed:
Router doc update

johanandren committed:
Call the routing logic also with an empty set of routees

johanandren committed:
Test fail fix

The group router would subscribe twice to the receptionist, once in the
initial state and once in the final state. The final one caused a race
in the test.

Also added a debug log entry when the router gets an update

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,65 @@ object Receptionist extends ExtensionId[Receptionist] {
+     * In a non-clustered `ActorSystem` this will be `false` for all listings.
+     * For `Find` queries and the initial listing for a `Subscribe` this will always be `false`.
+     */
+    def onlyReachabilityChanged: Boolean
```
I'm not sure this is a good name change. Nowhere else in the API signatures we have reachability. Could be good to not tie it too close to unreachable/reachable so that we can use it for preferred service instances based on other other health or load metrics in the future.

```
def isAllServiceInstancesChanged: Boolean
```

johanandren replied:
I think `isAllServiceInstancesChanged` is hard to parse though. What about `wasRegisteredActorsChanged`?

johanandren replied:
`wasServicesAddedOrRemoved`

patriknw replied:
that is a good name!

patriknw commented on akka-docs/src/main/paradox/typed/actor-discovery.md:
```diff
@@ -73,8 +73,15 @@ sends a `Ping` message and when receiving the `Pong` reply it stops.
+of the other nodes of the cluster.
 
-The state for the receptionist is propagated via @ref:[distributed data](../distributed-data.md) which means that each node will eventually reach the same set of actors per `ServiceKey`.
+The state for the receptionist is propagated via @ref:[distributed data](../distributed-data.md) which means that each node 
```
Distributed Data

johanandren committed:
Review feedback addressed

helena reviewed (approved):
Looks great

chbatey commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -340,6 +351,50 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
+          next(newRegistry)
+
+        case NodeReachable(uniqueAddress) =>
+          val keysForNode = registry.keysFor(uniqueAddress)
```
maybe extract method for this? Only difference is the log line

chbatey commented on akka-cluster-typed/src/multi-jvm/scala/akka/cluster/typed/internal/ClusterReceptionistUnreachabilitySpec.scala:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    "form a cluster" in {
+      runOn(first) {
```
Use MultiNodeTypedClusterSpec.formCluster

chbatey reviewed (approved):
LGTM with two comments

johanandren committed:
Review feedback applied

johanandren committed:
Merge branch 'wip-26355-take-cluster-unreachability-into-account' of github.com:johanandren/akka into wip-26355-take-cluster-unreachability-into-account

raboof commented on akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala:
```diff
@@ -269,15 +281,68 @@ object Receptionist extends ExtensionId[Receptionist] {
+     * In a non-clustered `ActorSystem` this will be `true` for all listings.
+     * For `Find` queries and the initial listing for a `Subscribe` this will always be `true`.
+     */
+    def servicesWereAddedOrRemoved: Boolean
```
:+1:
